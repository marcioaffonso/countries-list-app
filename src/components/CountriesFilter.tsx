import React from 'react';
import { View, Text, StyleSheet, TextInput } from 'react-native';
import PropTypes from 'prop-types';

export default function CountriesFilter({ onFilterChange }) {
  return (
    <View style={styles.filterContainer}>
      <Text style={styles.filterText}>Filter by name:</Text>
      <TextInput
        style={styles.filterInput}
        onChangeText={onFilterChange}
      />
    </View>
  );
};

CountriesFilter.propTypes = {
  onFilterChange: PropTypes.func,
};

const styles = StyleSheet.create({
  filterContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    margin: 10,
    marginTop: 30,
    minHeight: 30,
    maxHeight: 30,
    height: 30,
  },
  filterText: {
    flex: 1,
    fontSize: 20
  },
  filterInput: {
    flex: 1,
    borderRadius: 5,
    borderWidth: 1,
    backgroundColor: '#fff',
    borderColor: '#000',
    fontSize: 20,
    padding: 2
  },
});
