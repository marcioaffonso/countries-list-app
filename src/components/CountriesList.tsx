import React, { useDeferredValue } from 'react';
import { View, Text, FlatList, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

export default function CountriesList({ countries, onRefresh, refreshing }) {
  const deferredCustomers = useDeferredValue(countries);
  return (
    <FlatList
      data={deferredCustomers}
      renderItem={({ item }) => (
        <View style={styles.countryCard}>
          <View style={styles.nameContainer}>
            <Text style={styles.name}>{item.name.common}</Text>
            <Text style={styles.flag}>{item.flag}</Text>
          </View>
          <View style={styles.details}>
            <Text style={styles.detailsText}>üåé Name: { item.name.official }</Text>
            <Text style={styles.detailsText}>üè† Capital: { item.capital.join(', ') }</Text>
            <Text style={styles.detailsText}>
              üíµ Currency: { Object.keys(item.currencies).map((key) => {
                const currency = item.currencies[key];
                return `${key} (${currency.symbol}) - ${currency.name}`;
              }).join(', ') }
            </Text>
          </View>
        </View>
      )}
      ListEmptyComponent={() => (
        <View style={styles.countryCard}>
          <Text style={styles.detailsText}>No countries matching the filter criteria.</Text>
        </View>
      )}
      onRefresh={onRefresh}
      refreshing={refreshing}
    />
  );
};

CountriesList.propTypes = {
  countries: PropTypes.array,
  onRefresh: PropTypes.func,
  refreshing: PropTypes.bool
};

const styles = StyleSheet.create({
  countryCard: {
    backgroundColor: '#fff',
    borderRadius: 16,
    borderWidth: 2,
    margin: 10,
    padding: 16
  },
  nameContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 20
  },
  name: {
    fontSize: 30,
    fontWeight: 'bold',
    flexWrap: 'wrap'
  },
  flag: {
    fontSize: 30
  },
  details: {
    marginBottom: 10
  },
  detailsText: {
    fontSize: 18
  },
});
