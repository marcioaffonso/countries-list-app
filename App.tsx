import React from 'react';
import { SafeAreaView, StyleSheet, StatusBar } from 'react-native';
import { observer } from 'mobx-react-lite';
import CountriesFilter from './src/components/CountriesFilter';
import CountriesList from './src/components/CountriesList';
import countryStore from './src/stores/countryStore';
import AppLoader from './src/components/AppLoader';

const App = () => {
  return (
    <SafeAreaView style={styles.container}>
      {countryStore.loading && !countryStore.refreshing ?
        <AppLoader /> :
        <>
          <CountriesFilter onFilterChange={countryStore.setFilter} />
          <CountriesList
            countries={countryStore.filteredCountries}
            onRefresh={countryStore.loadCountries}
            refreshing={countryStore.refreshing}
          />
        </>
      }
    </SafeAreaView>
  );
};

export default observer(App);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#eee',
    paddingTop: StatusBar.currentHeight
  }
});
