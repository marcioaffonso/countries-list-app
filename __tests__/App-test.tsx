import React from 'react';
import renderer from 'react-test-renderer';
import { jest, expect, test } from '@jest/globals';
import App from '../App';
import CountriesFilter from '../src/components/CountriesFilter';
import CountriesList from '../src/components/CountriesList';

jest.mock('../src/stores/countryStore', () => {
  return {
    loading: false,
    filteredCountries: []
  };
});

test('renders correctly', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders App loading correctly', () => {
  const tree = renderer.create(<App />);
  expect(tree.root.findByType(CountriesFilter)).not.toBeNull;
  expect(tree.root.findByType(CountriesList)).not.toBeNull;
});
