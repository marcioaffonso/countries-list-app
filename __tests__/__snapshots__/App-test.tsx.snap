// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#eee",
      "flex": 1,
      "paddingTop": null,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 30,
        "justifyContent": "space-between",
        "margin": 10,
        "marginTop": 30,
        "maxHeight": 30,
        "minHeight": 30,
      }
    }
  >
    <Text
      style={
        {
          "flex": 1,
          "fontSize": 20,
        }
      }
    >
      Filter by name:
    </Text>
    <TextInput
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#000",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "fontSize": 20,
          "padding": 2,
        }
      }
    />
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        style={
          {
            "backgroundColor": "#fff",
            "borderRadius": 16,
            "borderWidth": 2,
            "margin": 10,
            "padding": 16,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
            }
          }
        >
          No countries matching the filter criteria.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
